src/ws.rs

```diff
@@ -1,78 +1,86 @@
-
-#[derive(PartialEq)]
-pub enum Events {
-    Json,
-    Lcds,
-    Callback,
-    None,
-}
+pub struct WSClient(WebSocketStream<MaybeTlsStream<TcpStream>>);
 
 impl WSClient {
-    pub async fn connect(events: Events) -> Result<Self, Error> {
-        let mut sys = System::new_all();
-        sys.refresh_all();
-        let process = find_process(&sys)?;
-        let args = extract_info(process)?;
-        let auth_token = encode_token(&args.0);
+    pub async fn connect() -> Result<Self, LcuWebsocketError> {
+        let (auth_token, port) = get_auth_info().map_err(|_| LcuWebsocketError::LcuNotAvailable)?;
 
-        let cert = native_tls::Certificate::from_pem(include_bytes!("./riotgames.pem"))?;
+        let cert = native_tls::Certificate::from_pem(include_bytes!("./riotgames.pem"))
+            .expect("invalid riotgames.pem certificate");
         let tls = native_tls::TlsConnector::builder()
             .add_root_certificate(cert)
-            .build()?;
-        let connector = tokio_tungstenite::Connector::NativeTls(tls);
-        let mut url = format!("wss://127.0.0.1:{}", args.1).into_client_request()?;
-        {
-            let headers = url.headers_mut();
-            headers.insert(
-                "Authorization",
-                HeaderValue::from_str(format!("Basic {}", auth_token).as_str())?,
-            );
-        }
+            .build()
+            .map_err(|_| LcuWebsocketError::AuthError)?;
+        let connector = Connector::NativeTls(tls);
 
-        let (ws_stream, _response) =
-            tokio_tungstenite::connect_async_tls_with_config(url, None, Some(connector)).await?;
+        let mut url = format!("wss://127.0.0.1:{port}")
+            .into_client_request()
+            .map_err(|_| LcuWebsocketError::AuthError)?;
+        url.headers_mut().insert(
+            "Authorization",
+            HeaderValue::from_str(format!("Basic {auth_token}").as_str())
+                .map_err(|_| LcuWebsocketError::AuthError)?,
+        );
 
-        let (mut write, read) = ws_stream.split();
-
-        match events {
-            Events::Json => write.send(Message::text("[5, \"OnJsonApiEvent\"]")).await?,
-            Events::Lcds => write.send(Message::text("[5, \"OnLcdsEvent\"]")).await?,
-            Events::Callback => write.send(Message::text("[5, \"OnCallback\"]")).await?,
-            Events::None => (),
-        }
+        let (ws_stream, _response) =
+            tokio_tungstenite::connect_async_tls_with_config(url, None, Some(connector))
+                .await
+                .map_err(|_| LcuWebsocketError::Disconnected)?;
 
-        Ok(Self { write, read })
+        Ok(Self(ws_stream))
     }
 
-    pub async fn subscribe(&mut self, event: String) -> Result<(), Error> {
-        self.write
-            .send(Message::text(format!("[5, {}]", event)))
-            .await?;
+    pub async fn subscribe(&mut self, event: LcuSubscriptionType) -> Result<(), LcuWebsocketError> {
```

Comment:
The code here is duplicated, and could be templated into something like
```rs
    pub async fn event(
        &mut self,
        event: LcuSubscriptionType,
        event_id: i8,
    ) -> Result<(), LcuWebsocketError> {
        self.0
            .send(Message::text(format!("[{event_id}, \"{event}\"]")))
            .await
            .map_err(|e| match e {
                Error::ConnectionClosed | Error::AlreadyClosed => LcuWebsocketError::Disconnected,
                _ => LcuWebsocketError::SendError,
            })
    }

Reply:
should we really provide an API for a variable event_id? As far as i know the only useful ids are 5/6 for subscribe/unsubscribe?

unless you mean something like this?

```rust
pub async fn subscribe(&mut self, subscription: LcuSubscriptionType) -> Result<(), LcuWebsocketError> {
        self.send_message(5, subscription).await
    }

pub async fn unsubscribe(
    &mut self,
    subscription: LcuSubscriptionType,
) -> Result<(), LcuWebsocketError> {
    self.send_message(6, subscription).await
}

async fn send_message(
    &mut self,
    event_id: u8,
    subscription: LcuSubscriptionType,
) -> Result<(), LcuWebsocketError> {
    self.0
        .send(Message::text(format!("[{event_id}, \"{subscription}\"]")))
        .await
        .map_err(|e| match e {
            Error::ConnectionClosed | Error::AlreadyClosed => {
                LcuWebsocketError::Disconnected(e.to_string())
            }
            _ => LcuWebsocketError::SendError,
        })
}
```

Reply:
Yes that is exactly what I mean